### ISSUE TITLE
HttpClient is created for each request which is anti-pattern

### ISSUE THREAD

**Issue by xperiandri** [fsprojects/FSharp.Data.GraphQL#275]
### Description

This line means that each property access creates on new `HttpClient`
https://github.com/fsprojects/FSharp.Data.GraphQL/blob/0f34f8fda9387f1fbba2294d8ff5c2f4e01f90bc/src/FSharp.Data.GraphQL.Client/GraphQLProviderRuntimeContext.fs#L20
Hence Dispose is called on a new instance.
And creating HttpClient for each request is an anti-pattern.

### Expected behavior

1. `HttpClient` is created once.
2. User can provide his instance of `HttpClient`

### Actual behavior

`HttpClient` is hardcoded and created on each property access.

### Related information 

* FSharp.Data.GraphQL 1.0.5

**Comment by johnberzy-bazinga** [comment]
Yeah, that's a good catch. That should be cached.

**Comment by xperiandri** [comment]
I hope to present a fix for tomorrow

**Comment by xperiandri** [comment]
As a temporary workaround, we can define a static mutable property on some module and if someone wants to swap `HttpClient` he will set that property.
Are you okay with such a solution?

**Comment by xperiandri** [comment]
@ivelten, what do you think?

**Comment by ivelten** [comment]
@xperiandri not sure if making it a static mutable property is the best approach as it will hit the functional style a bit. I think provide it statically, non-mutable as the default option could work (maybe internally making a singleton). And we could make an overload of the `GetContext` method of the provider to receive a custom `GraphQLClientConnection` and/or `HttpClient` if needed.

If you need help with this, I can make a pull request for the change this weekend.

**Comment by xperiandri** [comment]
@ivelten, yes, it would be extremely helpful!

I tried to implement that approach in 2 variants without any luck.
https://github.com/xperiandri/FSharp.Data.GraphQL/tree/http_client
https://github.com/xperiandri/FSharp.Data.GraphQL/tree/http_client_2
Could you help to determine what is wrong with my implementations?

**Comment by xperiandri** [comment]
My main goal is to be able to pass my HttpClient instance

**Comment by xperiandri** [comment]
@ivelten, have you tried my implementations?

**Comment by ivelten** [comment]
@xperiandri Sorry I did not see the answers you posted in time. I will try to develop a proposal tomorrow, since it's holiday in my country.

**Comment by ivelten** [comment]
I was searching a bit about `HttpClient`, more specifically about a Singleton/Caching usage of it, and I came to [this thread](https://stackoverflow.com/a/48778707). I guess we could avoid implementing a default instance using those approaches, as we could potentially run into DNS changes depending on the user. Instead, I just removed the HttpClient re-instantiation on each property access, and allowed the user to manage its own connection instance by passing a factory to an overload of `GetContext` method.

[Here is the PR](https://github.com/fsprojects/FSharp.Data.GraphQL/pull/285) with my proposal.
